### PR TITLE
Bznein/7023/on timeout packet

### PR DESCRIPTION
**Pull request by bznein**
## Description

closes: #7023

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against the correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#pull-request-targeting)).
- [ ] Linked to GitHub issue with discussion and accepted design, OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/build/building-modules/11-structure.md) and [Go style guide](../docs/dev/go-style-guide.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/main/testing/README.md#ibc-testing-package).
- [ ] Updated relevant documentation (`docs/`).
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Provide a [conventional commit message](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) to follow the repository standards.
- [ ] Include a descriptive changelog entry when appropriate. This may be left to the discretion of the PR reviewers. (e.g. chores should be omitted from changelog)
- [ ] Re-reviewed `Files changed` in the GitHub PR explorer.
- [ ] Review `SonarCloud Report` in the comment section below once CI passes.

### CODE REVIEW

File: modules/core/05-port/types/ibc_legacy_module.go [hunk 1]
```diff
@@ -269,12 +270,25 @@ func (im *LegacyIBCModule) OnAcknowledgementPacket(
...
 // OnTimeoutPacket implements the IBCModule interface
-func (LegacyIBCModule) OnTimeoutPacket(
+func (im *LegacyIBCModule) OnTimeoutPacket(
 	ctx sdk.Context,
 	channelVersion string,
 	packet channeltypes.Packet,
 	relayer sdk.AccAddress,
 ) error {
+	cbs := slices.Clone(im.cbs)
+	slices.Reverse(cbs)
```
Comment (chatton) [inline modules/core/05-port/types/ibc_legacy_module.go:280]: should be able to be removed after #7142

File: modules/core/keeper/msg_server.go [hunk 2]
```diff
@@ -532,11 +532,13 @@ func (k *Keeper) Timeout(goCtx context.Context, msg *channeltypes.MsgTimeout) (*
 		return nil, err
 	}
 
-	// Perform application logic callback
-	err = cbs.OnTimeoutPacket(ctx, channelVersion, msg.Packet, relayer)
-	if err != nil {
-		ctx.Logger().Error("timeout failed", "port-id", msg.Packet.SourcePort, "channel-id", msg.Packet.SourceChannel, "error", errorsmod.Wrap(err, "timeout packet callback failed"))
-		return nil, errorsmod.Wrap(err, "timeout packet callback failed")
+	for _, c := range cbs {
```
Comment (chatton) [inline modules/core/keeper/msg_server.go:535]: mega nit
```suggestion
	for _, cb := range cbs {
```

File: modules/core/keeper/msg_server.go [hunk 3]
```diff
@@ -532,11 +532,13 @@ func (k *Keeper) Timeout(goCtx context.Context, msg *channeltypes.MsgTimeout) (*
...
+		// Perform application logic callback
+		err = c.OnTimeoutPacket(ctx, channelVersion, msg.Packet, relayer)
```
Comment (chatton) [inline modules/core/keeper/msg_server.go:537]:
```suggestion
		err = cb.OnTimeoutPacket(ctx, channelVersion, msg.Packet, relayer)
```

File: modules/apps/callbacks/ibc_middleware_test.go [hunk 4]
```diff
@@ -550,8 +550,21 @@ func (s *CallbacksTestSuite) TestOnTimeoutPacket() {
...
+				var unmarshaller porttypes.PacketDataUnmarshaler
+				if legacyModule, ok := cbs[0].(*porttypes.LegacyIBCModule); ok {
+					legacyModuleCbs := legacyModule.GetCallbacks()
+					// transfer stack is at index 0
+					unmarshaller, ok = legacyModuleCbs[0].(porttypes.PacketDataUnmarshaler)
+					s.Require().True(ok)
+				} else {
+					unmarshaller, ok = cbs[0].(porttypes.PacketDataUnmarshaler)
+					s.Require().True(ok)
+				}
```
Comment (chatton) [inline modules/apps/callbacks/ibc_middleware_test.go:564]: I believed this should be removed in https://github.com/cosmos/ibc-go/issues/7083

Maybe we can add a link in the comment.

**Review by chatton — APPROVED** [review]
LGTM! just left a few small nits :rocket:

File: modules/core/05-port/types/ibc_legacy_module.go (hunk 1 above)
Comment (chatton) [reply modules/core/05-port/types/ibc_legacy_module.go:280]: either can be merged first, won't be a big deal to refactor

File: modules/apps/callbacks/ibc_middleware_test.go (hunk 4 above)
Comment (bznein) [reply modules/apps/callbacks/ibc_middleware_test.go:565]: Done!

**Pull request merged**